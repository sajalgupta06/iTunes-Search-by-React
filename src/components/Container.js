// @flow
import React from 'react';
import List from './List';
import Message from './Message';
import emitter from '../emitter';
import { getApiUrl } from '../utils';
import type { HeaderState, ContainerState } from '../type';

class Container extends React.PureComponent<{}, ContainerState> {
  state: ContainerState = {
    status: 'init',
    data: {}
  };

  async getSearchResult(headerState: HeaderState) {
    try {
      this.setState({ status: 'loading' });
      const resp = await fetch(getApiUrl(headerState));
      const json = await resp.json();
      this.setState({
        data: { ...json },
        status: json.resultCount ? '' : 'noContent'
      });
    } catch (e) {
      this.setState({ status: 'error' });
    }
  }

  componentDidMount() {
    emitter.on('search', this.getSearchResult.bind(this));
  }

  componentWillUnmount() {
    emitter.removeListener('search');
  }

  render() {
    const { status, data } = this.state;
    return (
      <div>
        {status.length ? <Message status={status} /> : <List {...data} />}
      </div>
    );
  }
}

export default Container;
