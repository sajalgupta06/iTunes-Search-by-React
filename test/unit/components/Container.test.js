import React from 'react';
import { shallow } from 'enzyme';
import Container from '../../../src/components/Container';

const mockData = {
  resultCount: 1,
  results: [
    {
      trackId: '1',
      trackPrice: 0.99,
      artworkUrl100: '',
      trackName: '',
      description: ''
    }
  ]
};

const mockResponse = {
  ok: true,
  json: _ => mockData
};

global.fetch = jest.fn().mockImplementation(_ => new Promise((resolve, reject) => resolve(mockResponse)));

describe('<Container />', () => {
  let component;
  beforeEach(() => {
    component = shallow(<Container />);
  });
  test('init state', () => {
    expect(component).toMatchSnapshot();
  });
  test('loading state', () => {
    component.setState({ status: 'loading' });
    expect(component).toMatchSnapshot();
  });
  test('show data', () => {
    const data = {
      resultCount: 2,
      results: [
        {
          trackName: 'test1'
        },
        {
          trackName: 'test2'
        }
      ]
    };
    component.setState({
      status: '',
      data
    });
    expect(component).toMatchSnapshot();
  });
  test('no search results state', () => {
    component.setState({ status: 'noContent' });
    expect(component).toMatchSnapshot();
  });
  test('error state', () => {
    component.setState({ status: 'error' });
    expect(component).toMatchSnapshot();
  });

  test('getSearchResult should fetch data', async () => {
    const state = {
      media: 'music',
      query: 'bsb'
    };
    const instance = component.instance();
    await instance.getSearchResult(state);
    expect(component.state()).toMatchSnapshot();
  });

  test('should set error state if there is a fetch error', async () => {
    const instance = component.instance();
    await instance.getSearchResult();
    expect(component.state()).toMatchSnapshot();
  });
});
