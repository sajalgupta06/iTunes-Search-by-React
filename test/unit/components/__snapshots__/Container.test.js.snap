// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Container /> error state 1`] = `
<div>
  <Message
    status="error"
  />
</div>
`;

exports[`<Container /> getSearchResult should fetch data 1`] = `
Object {
  "data": Object {
    "resultCount": 1,
    "results": Array [
      Object {
        "artworkUrl100": "",
        "description": "",
        "trackId": "1",
        "trackName": "",
        "trackPrice": 0.99,
      },
    ],
  },
  "status": "",
}
`;

exports[`<Container /> init state 1`] = `
<div>
  <Message
    status="init"
  />
</div>
`;

exports[`<Container /> loading state 1`] = `
<div>
  <Message
    status="loading"
  />
</div>
`;

exports[`<Container /> no search results state 1`] = `
<div>
  <Message
    status="noContent"
  />
</div>
`;

exports[`<Container /> should set error state if there is a fetch error 1`] = `
Object {
  "data": Object {},
  "status": "error",
}
`;

exports[`<Container /> show data 1`] = `
<div>
  <List
    resultCount={2}
    results={
      Array [
        Object {
          "trackName": "test1",
        },
        Object {
          "trackName": "test2",
        },
      ]
    }
  />
</div>
`;
